src/python/pants/rules/core/fmt.py

```diff
@@ -32,16 +32,33 @@ class Fmt(Goal):
   # Blocked on https://github.com/pantsbuild/pants/issues/8351
   name = 'fmt-v2'
 
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register('--transitive', type=bool, default=True,
+             help="If false, act only on the targets directly specified on the command line. "
+                  "If true, act on the transitive dependency closure of those targets.")
+
 
 @console_rule
-def fmt(console: Console, targets: HydratedTargets, union_membership: UnionMembership) -> Fmt:
+def fmt(
+  console: Console,
+  fmt_options: Fmt.Options,
+  transitive_targets: TransitiveHydratedTargets,
+  union_membership: UnionMembership
+) -> Fmt:
+
+  transitive = fmt_options.values.transitive
+
+  targets = transitive_targets.closure if transitive else transitive_targets.roots
```

Comment:
So, computing transitive targets and then only using the roots throws away quite a bit of effort in the case of `--no-transitive`.

To make this decision lazily, this rule should probably accept `Specs` as its argument (rather than `TransitiveHydratedTargets`), and then conditionally request `HydratedTargets` or `TransitiveHydratedTargets` depending on the option value.